Add after authenticate job option

alanhill opened the pull request:
### What

Adds the ability to generate an after authenticate job which can perform async or inline once the shop has authenticated with the app.

Branch renamed from: https://github.com/Shopify/shopify_app/pull/428 to better reflect the changes. 

@Hammadk @kevinhughes27 ready for review again - made the config a hash accepting `job` and `inline`.

kevinhughes27 commented on lib/generators/shopify_app/add_after_authenticate_job/add_after_authenticate_job_generator.rb:
```diff
@@ -0,0 +1,38 @@
+
+      def init_after_authenticate_config
+        initializer = load_initializer
+        return if initializer.include?("config.webhooks")
```
why do you need to return if they have have webhooks configured?

alanhill replied:
🤦‍♂️ typo

kevinhughes27 commented on lib/generators/shopify_app/add_after_authenticate_job/add_after_authenticate_job_generator.rb:
```diff
@@ -0,0 +1,38 @@
+
+        inject_into_file(
+          'config/initializers/shopify_app.rb',
+          "  config.after_authenticate_job = { job: Shopify::AfterAuthenticateJob, inline: false }\n",
```
I would generate this without the inline arg since most people should use it as a job

kevinhughes27 replied:
oh I see - its not really a default argument the way I was thinking of it. Maybe it is better to have here since otherwise it would only live in documentation. What are your thoughts @Hammadk

Hammadk replied:
I agree that we should keep it around for documentation purposes

Hammadk commented on lib/shopify_app/sessions_concern.rb:
```diff
@@ -87,5 +88,16 @@ def return_address
+    def perform_after_authenticate_job
+      return unless ShopifyApp.configuration.after_authenticate_job
+
+      job = ShopifyApp.configuration.after_authenticate_job[:job]
```
Is it possible for `ShopifyApp.configuration.after_authenticate_job` to be truthy, without setting `ShopifyApp.configuration.after_authenticate_job[:job]` ?

alanhill commented:
> Is it possible for ShopifyApp.configuration.after_authenticate_job to be truthy, without a job set in ShopifyApp.configuration.after_authenticate_job[:job]

Yes, I switched it up to check for the presence of the `config` in general, then `[:job]` and `[:inline]` otherwise it would fail on a nil class.

Also removed my typo and added the error handler like I originally intended. 

Ready for another quick round of 👀

Hammadk commented on lib/shopify_app/sessions_concern.rb:
```diff
@@ -87,5 +88,16 @@ def return_address
+
+      return unless config && config[:job].present? && config[:inline].present?
+
+      if config[:inline]
```
We should add tests for this conditional.

alanhill replied:
I was pondering on this for a bit. There doesn't seem to be anything testing the authentication flow of the app - installing script tags and webhooks got lucky because they are contained in their own class instead of a job firing from within the app itself and not the gem, therefore you're able to unit test vs. integration test. 

I think the only way to properly tests this is to have an integration test which goes through the entire authentication flow, unless there's a simpler way that I'm missing here?

Hammadk replied:
Have a look at the [`SessionControllerTests`](https://github.com/Shopify/shopify_app/blob/master/test/controllers/sessions_controller_test.rb#L97-L106). One way to do this is by testing the `#callback` method.

alanhill replied:
💡 ahhh that's what I was missing. TY.

alanhill commented:
Nice call on the test 😉 totally missed that one.

Should be good for review again!

Todo before :shipit: :

- [ ] Add to README
- [ ] Add to Changelog
- [ ] Bump version (minor I assume?)

kevinhughes27 commented on .byebug_history:
```diff
@@ -0,0 +1,5 @@
+config[:inline].present?
+config[:job].present?
+config
+next
```
🔥

alanhill replied:
🤦‍♂️

Hammadk commented on lib/shopify_app/sessions_concern.rb:
```diff
@@ -87,5 +88,16 @@ def return_address
+
+      return unless config && config[:job].present? && config[:inline].in?([true, false])
+
+      if config[:inline]
```
Do we have a default value for `config[:inline]` ? If not, we won't get to this statement because most people won't set ` && config[:inline].in?([true, false])`. 

```
return unless config && config[:job].present?

if config[:inline] == true
...
```

alanhill replied:
By default we add inline: false when you generate the job, but I see your point. So should we just consider anything but true to mean you want to perform asynchronously ?

Hammadk replied:
Yeah, `config[:inline] == true` is good because it's self documenting code. If someone thinking about running an inline job, reading this code would tell them that they need to set that value to `true`.

Hammadk commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -149,6 +155,51 @@ class SessionsControllerTest < ActionController::TestCase
+
+        mock_shopify_omniauth
+        get :callback, params: { shop: 'shop' }
+      end
```
Indentation is off here.

alanhill commented:
This should be ready for review - comments are fixed up, README and Changelog added, version bumped a minor version.

kevinhughes27 commented on README.md:
```diff
@@ -23,6 +23,7 @@ Table of Contents
  * [Home Controller Generator](#home-controller-generator)
  * [App Proxy Controller Generator](#app-proxy-controller-generator)
  * [Controllers, Routes and Views](#controllers-routes-and-views)
+ * [After Authenticate Job and Generator](#after-authenticate-job-an-generator)
```
I would probably drop `Generator` from the section heading

kevinhughes27 replied:
Also I think this should go after the Webhooks and ScriptTags manager sections

alanhill replied:
Missed this one, updated!

kevinhughes27 commented on README.md:
```diff
@@ -323,6 +324,18 @@ end
+
+You can generate an after authenticate job by running `shopify_app:add_after_authenticate_job` which will add a line in `shopify_app.rb` giving the option
+to run a job of your choosing either inline or asynchronously after the shop has authenticated. By default the job will run asynchronously unless `inline: false`
+is specified.
```
I would write this more like this:

If your app needs to perform specific actions after it is installed ShopifyApp can queue or run a job of your choosing (note that we already provide support for automatically creating Webhooks and Scripttags). To configure the after authenticate job update your initializer as follows:

```ruby
ShopifyApp.configure do |config|
  config.add_after_authenticate_job = { job: Shopify::AfterAuthenticateJob }
end
```

If you need the job to run synchronously add the `inline` flag:

```ruby
ShopifyApp.configure do |config|
  config.add_after_authenticate_job = { job: Shopify::AfterAuthenticateJob, inline: true }
end
```

We've also provided a generator which creates a skeleton job and updates the initializer for you:

```
bin/rails g shopify_app:add_after_authenticate_job
```

alanhill replied:
👍 Thanks for that, I'm awful with words.

alanhill commented:
README is updated with gooder ( 😉 ) words!

kevinhughes27 commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+7.3.1
```
I'd make this 7.4 since we added a feature

alanhill replied:
👍 bumped.

kevinhughes27 commented:
🚢
